[#1] PR 위한 브랜치 생성

hyeonga493 opened the pull request:
## PR Overview
- PR 위한 브랜치 생성(feature/start)
- 프로젝트 구조
- 설정 정보
- 기본 필요 문서 파일 초기화
<br>

## 상세 
다이어리 기능을 제공하는 URL 저장 및 관리 애플리케이션
1. 프로젝트 구조
`Core-Data` : 비즈니스 로직과 데이터 관리를 한 모듈로 통합
모델 정의, 데이터베이스 관리, 네트워크 API 호출 담당<br>
`API` : 외부 서버와 통신 및 인증 관련 기능을 통합
서버 API 인터페이스 정의, 네트워크 통신, 사용자 인증 처리를 담당<br>
`Utils` : 애플리케이션 전체에서 공통적으로 사용되는 유틸리티 및 헬퍼 함수 제공 
날짜 및 시간 처리
이미지 처리
네트워크 상태 체크, 로깅<br>
`UI` : 사용자 인터페이스와 상호작용 관리
액티비티 및 프래그먼트
뷰 모델
사용자 이벤트 처리
데이터 바인딩을 통한 UI 업데이트<br>

hyeonga493 committed:
[#1] PR위한 브랜치 생성(feature/start)

hyeonga493 committed:
[#1] 프로젝트 초기 세팅

hyeonga493 committed:
[#1] gitignore 파일 수정

f-lab-lyan reviewed (commented):
`.idea`는 일단은 필요없을 것 같습니다.

f-lab-lyan commented on .gitignore:
```diff
@@ -0,0 +1,18 @@
+
+### IntelliJ IDEA ###
+.idea/
+*.iml
```
이 파일들이 실제로 있는지 확인하고, 안생기면, 이 설정도 필요없을 것 같습니다.

f-lab-lyan commented on .gitignore:
```diff
@@ -0,0 +1,18 @@
+*.log
+
+### General ###
+*~
```
`.gitignore_global`에 넣으면 더 적절할 것 같네요.

f-lab-lyan commented on settings.gradle:
이 파일에 있는 설정들을 build.gradle로 옮기면 어떻게 되나요?

hyeonga493 replied:
### settings.gradle
멀티 프로젝트 빌드에서 프로젝트 설정과 구성을 관리하는 데 사용합니다.
멀티 프로젝트 빌드의 설정 또는 초기화에 관련된 플러그인을 적용합니다.
**프로젝트 전체에 걸친 플러그인을 적용**

### build.gradle
개별 프로젝트 또는 서브 프로젝트의 빌드 프로세스를 정의하는 데 사용합니다.
빌드 작업을 수행하는 데 필요한 플러그인을 적용합니다.
`Java` `Android` `Kotlin`
**컴파일, 테스트, 배포 등의 작업을 수행하는 플러그인**

f-lab-lyan commented on build.gradle:
나중에 필요할 때 추가하면 좋을 것 같네요.

f-lab-lyan commented on api/build.gradle:
```diff
@@ -0,0 +1,30 @@
+plugins {
```
버전이 있는 플러그인과 없는 플러그인의 차이는 무엇일까요?
https://plugins.gradle.org/plugin/io.spring.dependency-management

hyeonga493 replied:
### 버전이 없는 플러그인
**장점**
- 버전을 관리할 필요가 없어 설정이 단순함
- Gradle이 빌드 최신 버전의 플러그인을 자동으로 찾아 사용하므로 최신 기능과 버그 수정이 쉬움

**단점**
- 최신 버전을 자동으로 사용하여 호환되지 않는 변경이 적용되는 경우 문제가 발생할 수 있음
- 다른 시간에 같은 프로젝트를 빌드하는 경우 다른 결과가 도출될 수 있어 재현성이 떨어질 수 있음

### 버전이 있는 플러그인
**장점**
- 특정 버전을 명시함으로 항상 동일한 버전의 플러그인이 사용되어 일관성 있고 예측이 가능함
- 필수적인 플러그인의 특정 버전이 다른 프로젝트 구성 요소와 호환되도록 확실하게 할 수 있음
- 플러그인 업데이트로 인해 발생할 수 있는 문제를 예방할 수 있음
<br><br>

**써드파티 플러그인에 버전을 명시해야 하고 내장 플러그인에는 버전을 명시하지 않아도 오류가 발생하지 않습니다.**

### 내장 플러그인
Gradle 설치와 함께 제공되며 별도의 외부 저장소에서 다운로드 할 필요 없습니다.
Gradle이 업데이트될 때마다 해당 내장 플러그인들도 함께 업데이트 됩니다.

### 써드파티 플러그인
Gradle 플러그인 포털이나 다른 외부 소스에서 관리됩니다.
Gradle 버전과 독립적으로 버전이 관리되어 새로운 기능 추가나 버그 등의 이유로 지속적으로 업데이트 됩니다.
각 버전별로 기능이나 호환성에 차이가 있을 수 있어 특정 버전을 명시해야 합니다.

![image](https://github.com/f-lab-edu/hyeongarl/assets/127083504/8d5e205d-3e25-4248-8f6d-3d9c5d1a61a1)

f-lab-lyan commented on api/build.gradle:
```diff
@@ -0,0 +1,30 @@
+plugins {
+    id 'java'
+    id 'org.springframework.boot' version '3.2.5'
```
https://docs.gradle.org/current/dsl/org.gradle.plugin.use.PluginDependenciesSpec.html#N1B5AB

f-lab-lyan commented on api/build.gradle:
```diff
@@ -0,0 +1,30 @@
+version = '1.0-SNAPSHOT'
+
+java {
+    sourceCompatibility = '17'
```
https://docs.gradle.org/current/samples/sample_jvm_multi_project_with_toolchains.html#header

hyeonga493 replied:
Java 컴파일러가 코드를 처리하는 방법을 결정하는데 필요

`sourceCompatibility` : 소스 코드가 호환되도록 작성된 Java 버전을 지정합니다.
컴파일러에 소스 코드를 지정된 Java 버전용으로 작성된 것처럼 처리하도록 지시합니다.
코드에서 사용하는 Java 기능이 지정된 버전에서만 사용할 수 있도록 하여 코드가 실행될 수 있는 이전 JVM에서 인식되지 않는 최신 Java 버전의 기능을 실수로 사용하는 것을 방지합니다.

`targetCompatibility` : 컴파일러가 생성하는 Java 바이트 코드의 버전을 결정합니다. 애플리케이션을 실행할 것으로 예상되는 JVM의 버전입니다.
사용할 수 있는 Java 언어 기능을 제한하지만 출력 파일이 특정 JVM 버전과 호환되는지 확인합니다.
바이트코드 버전이 너무 최신이기 때문에 JVM이 코드를 이해하거나 실행할 수 없는 문제를 방지합니다.

`toolChain` : 개발 환경에 여러 JDK가 설치된 경우 프로젝트를 컴파일하고 실행하는 데 사용할 JDK를 정의할 수 있습니다.
```
java {
    toolchain {
        languageVersion = JavaLanguageVersion.of(17)
    }
}
```

f-lab-lyan commented on api/build.gradle:
```diff
@@ -0,0 +1,30 @@
+    implementation 'org.springframework.boot:spring-boot-starter-web-services'
+    testImplementation 'org.springframework.boot:spring-boot-starter-test'
+
+    // Lombok
```
https://plugins.gradle.org/plugin/io.freefair.lombok

hyeonga493 replied:
### dependencies 작성
```
dependencies {
    compileOnly 'org.projectlombok:lombok'
    annotationProcessor 'org.projectlombok:lombok'
}
```
**장점**
- 별도의 플러그인 설치 없이 Gradle 설정만으로 사용 가능
- 프로젝트에서 사용하는 모든 라이브러리를 확인할 수 있음

**단점**
- IDE와의 통합이나 다른 설정이 수동으로 필요할 수 있음(Lombok 플러그인 설치)
### plugins  작성
```
plugins {
  id "io.freefair.lombok" version "8.6"
}
```
**장점**
- 플러그인이 Lombok 관련 설정을 자동으로 처리하여 추가적인 설정을 자동화 할 수 있음
- 버전 관리와 설정이 쉬움

**단점**
- 프로젝트에 플러그인이 추가되어 플러그인의 유지보수 상태와 호환성을 주기적으로 확인해야 함

#### Lombok의 설정을 최소화하고 빠르게 구성하기 위해 plugin에 작성하여 플러그인을 사용하도록 변경

f-lab-lyan commented on api/build.gradle:
```diff
@@ -0,0 +1,30 @@
+}
+
+tasks.named('test') {
+    useJUnitPlatform()
```
이 설정은 왜 필요할까요?

hyeonga493 replied:
Junit5 테스트를 실행하기 위해 필요합니다.
JUnit5 는 `Junit Platform`, `Junit Jupiter`, `Junit Vintage` 의 모듈로 구성되어 있습니다.
`Junit Jupiter` : 새로운 테스트를 작성하기 위한 API 러너를 제공
`Junit Platform` : 플랫폼 러너를 통해 다양한 테스트 프레임워크를 실행할 수 있는 기반을 마련
`Junit Vintage` : 이전 버전의 JUnit으로 작성된 테스트를 실행할 수 있게 하는 모듈

기본적으로 Gradle은 Junit4를 사용하여 Java 테스트를 실행하므로 Junit5를 사용하려면 빌드 스트립트에 명시적으로 설정해야 합니다. 

`useJUnitPlatform()` 역할
1. 테스트 엔진 활성화
2. JUnit5 테스트 실행
3. 호환성 유지 `JUnit3`, `Junit4`

hyeonga493 committed:
[#1] .gitignore 불필요 코드 제거

hyeonga493 committed:
[#1] 불필요한 .idea 폴더 파일 제거

hyeonga493 committed:
[#1] new line 수정

hyeonga493 committed:
[#1] new line 수정

hyeonga493 committed:
[#1] build.gradle 수정

- Lombok dependencies 에서 plugins으로 변경
- Java 버전 sourceCompability 에서 toolchain 적용
